Sources/NIOTransportServices/NIOTSListenerChannel.swift

```diff
@@ -22,114 +22,55 @@ import Network
-    }
-
-    /// The parent `Channel` for this one, if any.
-    public let parent: Channel? = nil
-
-    /// The `EventLoop` this `Channel` belongs to.
-    internal let tsEventLoop: NIOTSEventLoop
-
-    private var _pipeline: ChannelPipeline! = nil  // this is really a constant (set in .init) but needs `self` to be constructed and therefore a `var`. Do not change as this needs to accessed from arbitrary threads.
-
-    internal let closePromise: EventLoopPromise<Void>
-
-    /// The underlying `NWListener` that this `Channel` wraps. This is only non-nil
-    /// after the initial connection attempt has been made.
-    private var nwListener: NWListener?
-
+internal final class NIOTSListenerChannel: StateManagedListenerChannel<NIOTSConnectionChannel> {
     /// The TCP options for this listener.
-    private let tcpOptions: NWProtocolTCP.Options
-
-    /// The TLS options for this listener.
-    private let tlsOptions: NWProtocolTLS.Options?
-
-    /// The `DispatchQueue` that socket events for this connection will be dispatched onto.
-    private let connectionQueue: DispatchQueue
-
-    /// An `EventLoopPromise` that will be succeeded or failed when a bind attempt succeeds or fails.
-    private var bindPromise: EventLoopPromise<Void>?
-
-    /// The state of this connection channel.
-    internal var state: ChannelState<ActiveSubstate> = .idle
-
-    /// The kinds of channel activation this channel supports
-    internal let supportedActivationType: ActivationType = .bind
-
-    /// The active state, used for safely reporting the channel state across threads.
-    internal var isActive0 = ManagedAtomic(false)
-
-    /// Whether a call to NWListener.receive has been made, but the completion
-    /// handler has not yet been invoked.
-    private var outstandingRead: Bool = false
-
-    /// Whether autoRead is enabled for this channel.
-    private var autoRead: Bool = true
-
-    /// The value of SO_REUSEADDR.
-    private var reuseAddress = false
-
-    /// The value of SO_REUSEPORT.
-    private var reusePort = false
-    
-    /// The value of the allowLocalEndpointReuse option.
-    private var allowLocalEndpointReuse = false
-
-    /// Whether to enable peer-to-peer connectivity when using Bonjour services.
-    private var enablePeerToPeer = false
-
-    /// The default multipath service type.
-    private var multipathServiceType = NWParameters.MultipathServiceType.disabled
-
-    /// The event loop group to use for child channels.
-    private let childLoopGroup: EventLoopGroup
+    private var tcpOptions: NWProtocolTCP.Options {
+        get {
+            guard case .tcp(let options) = protocolOptions else {
+                fatalError("NIOTSListenerChannel did not have a TCP protocol state")
+            }
 
-    /// The QoS to use for child channels.
-    private let childChannelQoS: DispatchQoS?
+            return options
+        }
+        set {
+            protocolOptions = .tcp(newValue)
```

Comment:
We should add a defensive check that we already had tcp options here, at least a debug mode one.